api/src/lib/expand-function/ast-helpers/ast-helpers.ts

```diff
@@ -0,0 +1,152 @@
+import fs from "node:fs";
+import * as ts from "typescript";
+import { URI } from "vscode-uri";
+import type { Definition as TSServerDefinition } from "../tsserver/index.js";
+import type { FunctionContextType } from "../types.js";
+
+export function getParentImportDeclaration(
+  node: ts.Node,
+): ts.ImportDeclaration | undefined {
+  let parent = node.parent;
+  while (parent && !ts.isSourceFile(parent)) {
+    if (ts.isImportDeclaration(parent)) {
+      return parent;
+    }
+    parent = parent.parent;
+  }
+  return undefined;
+}
+
+type DefinitionText = {
+  type: FunctionContextType;
+  text: string;
+  definitionNode?: ts.Node;
+};
+
+export function getDefinitionText(
+  node: ts.Node,
+  sourceFile: ts.SourceFile,
+): DefinitionText | undefined {
+  // Check: Variable declaration with initializer
+  // `const y = ...`
+  // `let x = ...`
+  if (ts.isVariableDeclaration(node) && node.initializer) {
+    return {
+      type: "unknown", // TODO
+      text: node.initializer.getText(sourceFile),
+    };
+  }
+
+  // Function declaration, arrow function, or function expression
+  // Declaration: `function f() { ... }`
+  // Arrow: `(c) => { ... }`
+  // Expression: `const h = function () { ... }` (or e.g., a function passed inline as a callback)
+  if (
+    ts.isFunctionDeclaration(node) ||
+    ts.isArrowFunction(node) ||
+    ts.isFunctionExpression(node)
+  ) {
+    return {
+      type: "function",
+      text: node.getText(sourceFile),
+      definitionNode: node,
+    };
+  }
+
+  // Check if the node is an identifier and if the parent node is a function declaration
+  if (ts.isIdentifier(node) && ts.isFunctionDeclaration(node.parent)) {
+    return {
+      type: "function",
+      text: node.parent.getText(sourceFile),
+      definitionNode: node.parent,
+    };
+  }
+
+  // Check if the node is an identifier and the parent node is a variable declaration with initializer
+  // `const myVariable = someValue;`
+  if (
+    ts.isIdentifier(node) &&
+    ts.isVariableDeclaration(node.parent) &&
+    node.parent.initializer
+  ) {
+    return {
+      type: "unknown", // TODO
+      text: node.parent.initializer.getText(sourceFile),
+    };
+  }
+
+  // Check for type alias declarations
+  // `type MyType = ...`
+  if (ts.isTypeAliasDeclaration(node)) {
+    return {
+      type: "type",
+      text: node.getText(sourceFile),
+      definitionNode: node,
+    };
+  }
+
+  // Check if the node is an identifier and the parent node is a type alias declaration
+  if (ts.isIdentifier(node) && ts.isTypeAliasDeclaration(node.parent)) {
+    return {
+      type: "type",
+      text: node.parent.getText(sourceFile),
+      definitionNode: node.parent,
+    };
+  }
+
+  return undefined;
+}
+
+export function findNodeAtPosition(
+  sourceFile: ts.SourceFile,
+  position: { line: number; character: number },
+): ts.Node | undefined {
+  function find(node: ts.Node): ts.Node | undefined {
```

Comment:
not really — the `ts.SourceFile` type is more specialized than a `ts.Node` as far as i understand. 

the sourceFile needs to be used to find the position of the node.

i suppose the function could accept an additional argument, `node`, then this could indeed be simplified.

Reply:
Right, having the node option would probably mean that we'd need to describe why there's an optional node parameter so this wouldn't really improve things. I just don't like nested code/functions, but this is fine.